pkgs/desktops/gnome/extensions/update-extensions.py

```diff
@@ -0,0 +1,291 @@
+package_name_registry: Dict[ShellVersion, Dict[PackageName, List[Uuid]]] = {}
+for shell_version in supported_versions.keys():
+    package_name_registry[shell_version] = {}
+
+
+def fetch_extension_data(uuid: str, version: str) -> Tuple[str, str]:
+    """
+    Download the extension and hash it. We use `nix-prefetch-url` for this for efficiency reasons
+    Returns a tuple with the hash (Nix-compatible) of the zip file and the hex-encoded content of its metadata.json.
+    """
+
+    # The download URLs follow this schema
+    uuid = uuid.replace("@", "")
+    url: str = f"https://extensions.gnome.org/extension-data/{uuid}.v{version}.shell-extension.zip"
+
+    # Yes, we download that file three times:
+
+    # The first time is for the maintainter, so they may have a personal backup to fix potential issues
+    #subprocess.run(
+    #    ["wget", url], capture_output=True, text=True
+    #)
+
+    # The second time, we extract the metadata.json because we need it too
+    with urllib.request.urlopen(url) as response:
+        data = zipfile.ZipFile(io.BytesIO(response.read()), 'r')
+        metadata = data.read('metadata.json').hex()
+
+    # The third time is to get the file into the store and to get its hash
+    hash = subprocess.run(
+        ["nix-prefetch-url", "--unpack", url], capture_output=True, text=True
+    ).stdout.strip()
+
+    return hash, metadata
+
+
+def generate_extension_versions(
+    extension_version_map: Dict[ShellVersion, ExtensionVersion], uuid: str
+) -> Dict[ShellVersion, Dict[str, str]]:
+    """
+    Takes in a mapping from shell versions to extension versions and transforms it the way we need it:
+    - Only take one extension version per GNOME Shell major version (as per `supported_versions`)
+    - Filter out versions that only support old GNOME versions
+    - Download the extension and hash it
+    """
+    extension_versions: Dict[ShellVersion, Dict[str, str]] = {}
+    for shell_version, version_prefix in supported_versions.items():
+        # Newest compatible extension version
+        extension_version: Optional[int] = max(
+            (
+                int(ext_ver)
+                for shell_ver, ext_ver in extension_version_map.items()
+                if (shell_ver.startswith(version_prefix))
+            ),
+            default=None,
+        )
+        # Extension is not compatible with this GNOME version
+        if not extension_version:
+            continue
+        logging.debug(
+            f"[{shell_version}] Downloading '{uuid}' v{extension_version}"
+        )
+        sha256, metadata = fetch_extension_data(uuid, str(extension_version))
+        extension_versions[shell_version] = {
+            "version": str(extension_version),
+            "sha256": sha256,
+            # The downloads are impure, their metadata.json may change at any time.
+            # Thus, be back it up / pin it to remain deterministic
```

Comment:
Adding a link to https://gitlab.gnome.org/Infrastructure/extensions-web/issues/137 to the comment would be helpful too.